src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -101,19 +101,31 @@ public class SqlMutationEngine : IMutationEngine
             // If authorization fails, an exception will be thrown and request execution halts.
             AuthorizeMutationFields(context, parameters, entityName, mutationOperation);
 
+            string roleName = GetHttpContext().Request.Headers[AuthorizationResolver.CLIENT_ROLE_HEADER];
+
+            // The presence of READ permission is checked in the current role (with which the request is executed) as well as Anonymous role. This is because, for GraphQL requests,
+            // READ permission is inherited by other roles from Anonymous role when present.
+            bool isReadPermissionConfigured = _authorizationResolver.AreRoleAndOperationDefinedForEntity(entityName, roleName, EntityActionOperation.Read)
+                                              || _authorizationResolver.AreRoleAndOperationDefinedForEntity(entityName, AuthorizationType.Anonymous.ToString(), EntityActionOperation.Read);
+
```

Comment:
read operation permissions checks are performed by hot chocolate by observing the `@authorize` directives present on the object type definition and/or its fields. 

so for my understanding, you just have this check here to determine whether to make follow-up roundtrip to the db to get the inserted data?

You'll also need to consider the case when an entity has the "read" permission defined, but excludes all fields. probably another opportunity to break this into a private helper.

Reply:
>read operation permissions checks are performed by hot chocolate by observing the @authorize directives present on the object type definition and/or its fields. so for my understanding, you just have this check here to determine whether to make follow-up roundtrip to the db to get the inserted data?

True, but HotChocolate performs this check when constructing the response (after all the database operations have taken place). But, the `select` query doesn't have to be executed when it can be determined that an error response is going to be returned anyways.  Hence, added this check before the select query execution.

> You'll also need to consider the case when an entity has the "read" permission defined, but excludes all fields. probably another opportunity to break this into a private helper.


It looks the execution of `select` query can be skipped for more scenarios. For example, in this mutation, I've excluded only `publisher_id` field. But, the selection set contains `publisher_id`. Even, in cases such as this, we don't return any data (for other valid fields). 
![image](https://github.com/Azure/data-api-builder/assets/11196553/ecb66abc-0973-4d4f-999d-9eb3dd63827f)

So, even there is one field in the selection set that is not queryable, the select query execution can be skipped as no data for any of the fields is returned. I'm looking into adding this logic change

Reply:
>  looks the execution of select query can be skipped for more scenarios. For example, in this mutation, I've excluded only publisher_id field. But, the selection set contains publisher_id. Even, in cases such as this, we don't return any data (for other valid fields). [This was performed using version 0.9.7 ]

Can you include the config you used here? Also, can you try this where no anonymous role is defined? only authenticated and other custom roles. Is the behavior the same?

Reply:
Yes, the behavior is same when anonymous is not defined. 

```json
"Book": {
      "source": {
        "object": "books",
        "type": "table"
      },
      "graphql": {
        "enabled": true,
        "type": {
          "singular": "book",
          "plural": "books"
        }
      },
      "rest": {
        "enabled": true
      },
      "permissions": [
        {
          "role": "authenticated",
          "actions": [
            {
              "action": "create"
            },
            {
              "action": "update"
            },
            {
              "action": "delete"
            },
            {
               "action": "read",
               "fields": {
                   "exclude": [ "publisher_id" ] 
                }  
            }
          ]
        }
   ]
   ....
 }
```